### ISSUE TITLE
sum(::Tridiagonal, dims = 2) is slow

### ISSUE THREAD

**Issue by matthieugomez** [JuliaLang/julia#34553]
Reducing rows of a Tridiagonal matrix is much slower than reducing its columns. This is surprising since its storage is not column-based.

```julia
using BenchmarkTools
N = 1000
T = Tridiagonal(ones(N-1), ones(N), ones(N-1))
@btime sum(T, dims =1)
#   1.368 ms (24 allocations: 40.86 KiB)
@btime sum(T, dims =2)
#    41.458 ms (30 allocations: 33.14 KiB)
```

**Comment by andreasnoack** [comment]
I think that at least part of the problem is that the return type is a `SparseMatrixCSC`. The index operations for a sparse 1000x1 matrix just happens to be much more expensive than the 1x1000 case. Maybe we should add specialized versions of some of the reductions for the special diagonal types. E.g. we could also have pretty fast versions of `prod` for these.

**Comment by StefanKarpinski** [comment]
Summing a Diagonal matrix should produce a dense column or row matrix, no? It should just copy the diagonal vector into whatever the correct shape is which should be very fast.

**Comment by matthieugomez** [comment]
Yes I think tridiagonal should also produce a dense column.

**Comment by matthieugomez** [comment]
In the meantime, one can just to the following:
```julia
 @btime sum(T', dims = 1)
#   1.368 ms (24 allocations: 40.86 KiB)
```

**Comment by ranocha** [comment]
I think this can be closed after #35200 has been merged. Now, I get
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> N = 1_000; T = Tridiagonal(ones(N-1), ones(N), ones(N-1));

julia> @benchmark sum(T, dims=1)
BenchmarkTools.Trial: 
  memory estimate:  7.95 KiB
  allocs estimate:  2
  --------------
  minimum time:     664.087 ns (0.00% GC)
  median time:      1.186 μs (0.00% GC)
  mean time:        1.394 μs (7.61% GC)
  maximum time:     11.258 μs (85.11% GC)
  --------------
  samples:          10000
  evals/sample:     103

julia> @benchmark sum(T, dims=2)
BenchmarkTools.Trial: 
  memory estimate:  8.14 KiB
  allocs estimate:  8
  --------------
  minimum time:     1.002 μs (0.00% GC)
  median time:      1.520 μs (0.00% GC)
  mean time:        1.777 μs (7.19% GC)
  maximum time:     88.536 μs (94.32% GC)
  --------------
  samples:          10000
  evals/sample:     10
```
on my notebook. That's three orders of magnitude faster than the choice `dims=1` before.

**Comment by matthieugomez** [comment]
Awesome! Many thanks @ranocha
